fields/types/gcsimage/GcsImageType.js

```diff
@@ -279,13 +279,16 @@ gcsimage.prototype.uploadFile = function (item, file, update, callback) {
     }).then(function () {
       const dimensions = sizeOf.sync(fs.readFileSync(file.path));
 
+      const url = isPublicRead ? gcsHelper.getPublicUrl(_this.options.bucket, gcsDir + filename) :
+        gcsHelper.getAuthenticatedUrl(_this.options.bucket, gcsDir + filename);
```

Comment:
If we are overwriting the data ultimately, why do we want to build the URL here?

Reply:
`@twreporter/keystone` is a library.
`@twreporter/plate` uses `@twreporter/keystone` library to build the server.

For library perspective, we need to provide the URL to indicate the users where the photo are uploaded.

Reply:
> `@twreporter/keystone` is a library.
> `@twreporter/plate` uses `@twreporter/keystone` library to build the server.
> 
> For library perspective, we need to provide the URL to indicate the users where the photo are uploaded.
> 

I'm second to build the raw photo URL through `@twreporter/keystone`.  The problem I proposed here was to address the raw photo `url` stored in the mongo is overwritten to the resized mobile in plate. This does not make sense from the data perspective. Also, if that url is only for display on keystone, do we really need to store these information in the mongodb?

> However, for us (`@twreporter/plate`), we upload the photo onto a bucket which is not public, therefore, we need to overwrite the URL to render on `https://keystone-editor.twreporter.org/keystone/images/${ObjectID}`.

This is the major problem we should address: `separate the display from the `url` field.`

If that could be possible, we could generate the address dynamic during the runtime with [signed url](https://cloud.google.com/storage/docs/access-control/signed-urls) for listing within a period of time

Reply:
> The problem I proposed here was to address the raw photo url stored in the mongo is overwritten to the resized mobile in plate. This does not make sense from the data perspective. Also, if that url is only for display on keystone, do we really need to store these information in the mongodb?

This `url` property is used by `fields/types/gcsimage/GcsImageField.js`.
Since `GcsIMageField` is using `url` property to display that photo, we have to store the photo in that specific data structure.

For keystone gcsimage type perspective, it uploads the photo onto GCS, and it needs to show that photo to end users as well. Therefore, it must store either public URL, or authenticated URL or `gs://` URI.
For our case, we don't make our bucket `twreporter-raw-photos` public; hence, we could only provide authenticated URL.
However, it's useless if plate provide authenticated URL to end users.
End users will not see the uploaded photos at all.

Thus, we replace that authenticated URL by mobile URL to show the uploaded photos to end users.

> This is the major problem we should address: separate the display from the url field.

We need to modify the keystone a lot to make this happen.

For what I know about KeystoneJS, it binds data models with displaying.
Therefore, in every type, such as `gcsimage`, in `fields/types/gcsimage` folder,
you can see `GcsImageField.js`,  `GcsImageType.js`, `GcsImageColumn.js` and `GcsImageFilter.js`

Reply:
Current image list page already displays the image column with image in the mobile resolution and I found this [PR](https://github.com/twreporter/keystone/commit/d2c02327e7816eaad9352ddc62855b1661b5b8b6#diff-afcd87ed575a4814179b000f51e539225d856dcc644c7a53d8a05ccefbafaa21). Do we need to tweak this as well?

Reply:
> Current image list page already displays the image column with image in the mobile resolution and I found this [PR](https://github.com/twreporter/keystone/commit/d2c02327e7816eaad9352ddc62855b1661b5b8b6#diff-afcd87ed575a4814179b000f51e539225d856dcc644c7a53d8a05ccefbafaa21). Do we need to tweak this as well?

Yes, image column uses mobile URL to display the photos.
Hence, I prefer to use mobile URL as well at image field page (`https://keystone-editor.twreporter.org/keystone/images/${ObjectID}`).

Reply:
Thanks for reminder.
If we accept to store mobile image URL into `url` property, we could also update `GcsImageColumn.js` to get image url from `url` property, rather than `resizedTargets.mobile.url` (see https://github.com/twreporter/keystone/blob/master/fields/types/gcsimage/GcsImageColumn.js#L11) .

Reply:
The another option is that we rewrite `GcsImageField.js` (the cost will be high).

Therefore, we could redesign how we display the uploaded photos.

Reply:
> Thanks for reminder.
> If we accept to store mobile image URL into `url` property, we could also update `GcsImageColumn.js` to get image url from `url` property, rather than `resizedTargets.mobile.url` (see https://github.com/twreporter/keystone/blob/master/fields/types/gcsimage/GcsImageColumn.js#L11) .

If the `GCSImageColumn` is rendered with mobile URL, could you explain why we still require to overwrite the URL?

Reply:
`GcsImageField.js` extends from `fields/types/cloudinaryimage/CloudinaryImageField.js`.
In `CloudinaryImageField.js`, it uses `url` property to  display the photo.


There are tradeoffs we have to make:
### 1. Replace authenticated URL by mobile URL at `hooks.postUpload`:
#### Advantage: 
We don't have to update `GcsImageField.js`. For me, it's not easy to update that codes.

#### Disadvantage:
It's confusing why we replace authenticated URL by mobile URL in `plate/models/Image.js`.

### 2. Separate data models from displaying
#### Advantage:
The codes might be cleaner, and not decoupled with data structure we store in the mongodb.

#### Disadvantage:
It requires lots of effort to do so, and I am not quite confident of doing it.

Reply:
I will prefer to the second method.

I found this [PR](https://github.com/twreporter/keystone/commit/c7bcb6df6620590351f34b776e22065220fc83b1#diff-452520a2c99905137a81f9c2f3bf5060a0ddecde14d7fc09728d160d2dbb688b) from the history of GcsImageField.js.
I suggest we could attempt to duplicate the `CloudinaryImageField.js` rather than extend it directly. Then, tweak it to fit our own need and remove the unnecessary support. I believe [`getImageSource`](https://github.com/twreporter/keystone/blob/master/fields/types/cloudinaryimage/CloudinaryImageField.js#L68) and [`getImageURL`](https://github.com/twreporter/keystone/blob/master/fields/types/cloudinaryimage/CloudinaryImageField.js#L78) would be a good starting point.